pandas/plotting/_matplotlib/core.py

```diff
@@ -378,6 +394,21 @@ def _compute_plot_data(self):
                 label = "None"
             data = data.to_frame(name=label)
 
+        # GH15079 restructure data if by is defined
+        if self.by is not None:
+            self.subplots = True
+            grouped = data.groupby(self.by)
+            self._grouped_data_size = len(grouped)
+
+            data_list = []
+            for key, group in grouped:
+                columns = MultiIndex.from_product([[key], self.column])
+                sub_group = group[self.column]
+                sub_group.columns = columns
+                data_list.append(sub_group)
+
+            data = concat(data_list, axis=1)
```

Comment:
I know in all the other transformations inside this function it's done inline, and you're being consistent. But this is somehow complex, I think it should be in a separate function, so it's clear what's the input, what's the output, we can add a small docstring explaining and with an example. And most importantly, it'll be easier to test properly.

Reply:
yeah, okay, I moved it out as an internal function to transform the data. However, i found it hard to test, so I added a couple inline comments and docstrings, please let me know if there is a good way to test this function out! @datapythonista